src/dmd/semantic3.d

```diff
@@ -822,7 +822,17 @@ private extern(C++) final class Semantic3Visitor : Visitor
                             else if (exp.type.wildOf().implicitConvTo(tret))
                                 exp = exp.castTo(sc2, exp.type.wildOf());
                         }
-                        exp = exp.implicitCastTo(sc2, tret);
+
+                        bool hasCopyCtor;
+                        if (exp.type.ty == Tstruct)
+                        {
+                            TypeStruct ts = cast(TypeStruct)(exp.type);
+                            if (ts.sym.copyCtor)
+                                hasCopyCtor = true;
+                        }
```

Comment:
This can be written as a one-liner:
```
const bool hasCopyCtor = exp.type.ty == Tstruct &&
                                 (cast(TypeStruct)exp.type).sym.copyCtor !is null;
```
```